### PR TITLE
wip-fancy-striping

### PR DESCRIPTION
**Pull request by idryomov**

### CODE REVIEW

File: drivers/block/rbd.c [hunk 1]
```diff
@@ -5269,7 +5271,17 @@ static int rbd_dev_v2_header_onetime(struct rbd_device *rbd_dev)
 			goto out_err;
 	}
 
-	rbd_init_layout(rbd_dev);
+	ret = rbd_init_layout(rbd_dev);
+#if 0 /* TODO(dis) */
```
Comment (dillaman) [inline drivers/block/rbd.c]: Nit: remove reminder

File: drivers/block/rbd.c [hunk 2]
```diff
@@ -2040,6 +2039,158 @@ static int rbd_img_request_fill(struct rbd_img_request *img_request,
...
+static int rbd_img_fill_from_bio(struct rbd_img_request *img_req,
+				 u64 off, u64 len, struct bio *bio)
+{
+	struct ceph_file_extent ex = { off, len };
+	struct ceph_bio_iter it = { .bio = bio, .iter = bio->bi_iter };
+
+	return __rbd_img_fill_from_bio(img_req, &ex, 1, &it);
+}
+
+static void set_pos(struct ceph_object_extent *ex, u32 bytes, void *arg)
```
Comment (dillaman) [inline drivers/block/rbd.c]: Nit: ```set_bvec_pos``` for consistency w/ ```set_bio_pos```?

File: drivers/block/rbd.c [hunk 3]
```diff
@@ -2251,6 +2268,211 @@ static void rbd_img_obj_request_fill(struct rbd_obj_request *obj_request,
...
+		obj_req->write_state = RBD_OBJ_WRITE_FLAT;
+		num_osd_ops = 2; /* setallochint + write/writefull */
+	}
+
+	obj_req->osd_req = rbd_osd_req_create(rbd_dev, OBJ_OP_WRITE,
+					      num_osd_ops, obj_req);
+	if (!obj_req->osd_req)
+		return -ENOMEM;
+
+	if (obj_request_overlaps_parent(obj_req)) {
```
[the remark this answers is not in the document]
Comment (dillaman) [reply drivers/block/rbd.c]: ... never mind, I see these are all removed in a future commit

File: drivers/block/rbd.c (hunk 1 above)
Comment (idryomov) [reply drivers/block/rbd.c]: This one is actually something I wanted to talk about.  I'll bring it up tomorrow.

File: drivers/block/rbd.c [hunk 4]
```diff
@@ -2712,16 +2922,171 @@ static int rbd_img_request_submit(struct rbd_img_request *img_request)
...
+	}
+
+	osd_req_op_cls_init(obj_req->osd_req, 0, CEPH_OSD_OP_CALL, "rbd",
+			    "copyup");
+	osd_req_op_cls_request_data_bvecs(obj_req->osd_req, 0,
+					  obj_req->copyup_bvecs, bytes);
+
+	switch (rbd_img_request_op_type(obj_req->img_request)) {
+	case OBJ_OP_WRITE:
+		__rbd_obj_setup_write(obj_req, 1);
```
[the remark this answers is not in the document]
Comment (idryomov) [reply drivers/block/rbd.c]: It writes to the new OSD request, so it doesn't overwrite anything.  The "stat + ops" OSD request is destroyed a few lines above and a new "copyup + ops" OSD request is created.  "1" is to skip over the copyup op.

File: drivers/block/rbd.c [hunk 5]
```diff
@@ -1831,8 +1850,6 @@ static struct ceph_osd_request *rbd_osd_req_create(
 		snapc = img_request->snapc;
 	}
 
-	rbd_assert(num_ops == 1 || ((op_type == OBJ_OP_WRITE) && num_ops == 2));
-
 	return __rbd_osd_req_create(rbd_dev, snapc, num_ops,
 	    (op_type == OBJ_OP_WRITE || op_type == OBJ_OP_DISCARD) ?
```
[the remark this answers is not in the document]
Comment (idryomov) [reply drivers/block/rbd.c]: This piece of code is removed in "rbd: simplify rbd_osd_req_create()".

File: drivers/block/rbd.c [hunk 6]
```diff
@@ -2670,22 +2668,19 @@ static int rbd_img_obj_exists_submit(struct rbd_obj_request *obj_request)
 	 *         le32 tv_nsec;
 	 *     } mtime;
 	 */
-	size = sizeof (__le64) + sizeof (__le32) + sizeof (__le32);
-	page_count = (u32)calc_pages_for(0, size);
-	pages = ceph_alloc_page_vector(page_count, GFP_NOIO);
+	pages = ceph_alloc_page_vector(1, GFP_NOIO);
```
[the remark this answers is not in the document]
Comment (idryomov) [reply drivers/block/rbd.c]: Correct.  It's the last parameter of osd_req_op_raw_data_in_pages() call a couple of lines below.  If it's true, the OSD client takes care of releasing the pages when it's done with the message.

It's on my list to get rid of this allocation altogether -- we only look at the error code, not the actual stat response.

File: include/linux/ceph/messenger.h [hunk 7]
```diff
@@ -139,6 +141,42 @@ struct ceph_bio_iter {
 
 #endif /* CONFIG_BLOCK */
 
+struct ceph_bvec_iter {
+	struct bio_vec *bvecs;
+	struct bvec_iter iter;
+};
+
+#define __ceph_bvec_iter_advance_step(it, n, STEP) do {			      \
```
[the remark this answers is not in the document]
Comment (idryomov) [reply include/linux/ceph/messenger.h]: That's the single space tab thing.

File: drivers/block/rbd.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (idryomov) [reply drivers/block/rbd.c]: Yeah, it's better with rbd_obj_calc_img_extents().  Given that parent_overlap can change at any time (e.g. when the request is in flight), I'm not sure if we can ever achieve perfection here...

File: drivers/block/rbd.c [hunk 8]
```diff
@@ -217,12 +217,14 @@ struct rbd_obj_request;
 typedef void (*rbd_obj_callback_t)(struct rbd_obj_request *);
 
 enum obj_request_type {
-	OBJ_REQUEST_NODATA, OBJ_REQUEST_BIO, OBJ_REQUEST_PAGES
+	OBJ_REQUEST_NODATA = 1,
```
Comment (yangdongsheng) [inline drivers/block/rbd.c]: Hi @idryomov could you help to explain more about the background of this change? In my opinion, enum starts from 0 is good for array. for example:
```
enum operation_type {
          TYPE_ZERO = 0,
          TYPE_ONE,
          TYPE_TWO,
          TYPE_MAX,
};

int operation_counts[TYPE_MAX];

int show_operation_count(enum  operation_type) {
      if (operation_type >= TYPE_MAX)
            return -EINVAL;
      return operation_counts[operation_type]; 
}
```
Thanx

Comment (idryomov) [reply drivers/block/rbd.c]: Well, we don't have obj_req arrays indexed by OBJ_REQUEST_*.  It's used as a type identifier, and starting at 1 makes it easier to spot uninitialized use when prototyping or debugging.

Comment (yangdongsheng) [reply drivers/block/rbd.c]: okey, makes sense.

**Pull request closed without merging**